server/src/main/java/org/apache/iotdb/db/query/dataset/AlignByDeviceDataSet.java

```diff
@@ -116,25 +119,29 @@ protected boolean hasNextWithoutConstraint() throws IOException {
 
     while (deviceIterator.hasNext()) {
       currentDevice = deviceIterator.next();
-      Set<String> measurementColumnsOfGivenDevice = deviceToMeasurementsMap
-          .get(currentDevice);
-      executeColumns = new ArrayList<>(measurementColumnsOfGivenDevice);
-
-      // extract paths and aggregations if exist from executeColumns
+      // get all measurements of current device
+      Set<String> measurementOfGivenDevice = currentDevice.getChildren().keySet();
+      // extract paths and aggregations queried from all measurements
+      // executeColumns is for calculating rowRecord
+      executeColumns = new ArrayList<>();
       List<Path> executePaths = new ArrayList<>();
       List<TSDataType> tsDataTypes = new ArrayList<>();
       List<String> executeAggregations = new ArrayList<>();
-      for (String column : executeColumns) {
+      for (String column : measurementDataTpeMap.keySet()) {
         if (dataSetType == DataSetType.GROUPBY || dataSetType == DataSetType.AGGREGATE) {
-          Path path = new Path(currentDevice,
-              column.substring(column.indexOf('(') + 1, column.indexOf(')')));
-          tsDataTypes.add(tsDataTypeMap.get(path));
-          executePaths.add(path);
-          executeAggregations.add(column.substring(0, column.indexOf('(')));
+          String measurement = column.substring(column.indexOf('(') + 1, column.indexOf(')'));
+          if (measurementOfGivenDevice.contains(measurement)){
+            executeColumns.add(column);
+            executePaths.add(new Path(currentDevice.getFullPath(), measurement));
+            tsDataTypes.add(measurementDataTpeMap.get(column));
+            executeAggregations.add(column.substring(0, column.indexOf('(')));
+          }
         } else {
-          Path path = new Path(currentDevice, column);
-          tsDataTypes.add(tsDataTypeMap.get(path));
-          executePaths.add(path);
+          if (measurementOfGivenDevice.contains(column)) {
+            executeColumns.add(column);
+            executePaths.add(new Path(currentDevice.getFullPath(), column));
+            tsDataTypes.add(measurementDataTpeMap.get(column));
+          }
```

Comment:
try to extract the same process to avoid duplicated codes:
```
String measurement = column;
        if (dataSetType == DataSetType.GROUPBY || dataSetType == DataSetType.AGGREGATE) {
          measurement = column.substring(column.indexOf('(') + 1, column.indexOf(')'));
          if (measurementOfGivenDevice.contains(measurement)) {
            executeAggregations.add(column.substring(0, column.indexOf('(')));
          }
        }
        if (measurementOfGivenDevice.contains(measurement)) {
          executeColumns.add(measurement);
          executePaths.add(new Path(currentDevice.getFullPath(), measurement));
          tsDataTypes.add(measurementDataTpeMap.get(measurement));
        }
```
(please first ensure codes above work : ) )

Reply:
Thank you! It's very helpful for me.